pkg/raft/log_unstable.go

```diff
@@ -227,23 +227,26 @@ func (u *unstable) stableSnapTo(i uint64) {
 // restore resets the state to the given snapshot. It effectively truncates the
 // log after s.lastEntryID(), so the caller must ensure this is safe.
 func (u *unstable) restore(s snapshot) bool {
-	// All logs >= s.term are consistent with the log covered by the snapshot. If
-	// our log is such, disallow restoring a snapshot at indices below our last
+	mark := s.commitMark()
+	// All logs >= mark.Term are consistent with the log covered by the snapshot.
+	// If our log is such, disallow restoring a snapshot at indices below our last
 	// index, to avoid truncating a meaningful suffix of the log and losing its
 	// durability which could have been used to commit entries in this suffix.
 	//
 	// In this case, the caller must have advanced the commit index instead.
 	// Alternatively, we could retain a suffix of the log instead of truncating
 	// it, but at the time of writing the whole stack (including storage) is
 	// written such that a snapshot always clears the log.
-	if s.term <= u.term && s.lastIndex() < u.lastIndex() {
+	if mark.Term <= u.term && mark.Index < u.lastIndex() {
```

Comment:
> 1. the snapshot contains a (Index, Term) pair. This pair can be understood as an entryID - the snapshot is the materialized state of the entry. This can also be understood as a commit mark - the entry was committed (or it couldn't be materialized in a snapshot)

Yes, the `(Index,Term)` is an `entryID`, but not a `CommitMark` in this terminology. Since it's number 1 in your list, see if it changes the chain of thought.

See the [snapshot](https://github.com/cockroachdb/cockroach/blob/645eb8c99796b3b88f5631aa0fc92a011010ce64/pkg/raft/types.go#L165-L190) wrapper which encompasses both the pb snapshot (`entryID`) and the term of its sender. For notation simplicitly, let's say that a snapshot (or any committed log prefix) is `entryID`@`term`.

Example committed log: `[2 2 2 2 2 3 3 3] @ term 3`. Say entries `[2 2 2 2 2]` were not committed under term 2, and only became committed under term 3 (when it first happened that raft's commit criterion "fired"). If so, any `CommitMark` that covers these `[2 ... 2]` entries will have `CommitMark.Term >= 3`; and any leader log at term >= `CommitMark.Term` will have all these entries (or any acceptor log at `accTerm` >= `CommitMark.Term`, which is the same thing). The existence of a `CommitMark` tells a partial story how this entry got committed, and communicates provenance.

> 2. for better or worse, we erase the raft log when applying a snapshot. So we must never apply one that would have us remove any log entries that the local raft instance may have acked as part of a quorum.

Yep, the challenge in raft is that the acceptor's log and the learner's state (snapshot + committed log) are intertwined, particularly we reason about "accepted state" of this combined node (in voting etc) as a `f(acceptor, learner)`. So we should somehow combine their coordinate systems to work together correctly.

Right now, I think, this combined "system" is (`Term`, `lastEntryID`, `commitIndex`). And this PR moves things a bit in the direction of some decoupling. Clearly needs more thought and scrutiny.

> 3. if the snapshot commit mark's term is (strictly) leading the unstable term (which is the entryID.term of the newest (stable or unstable) entry, the snapshot commits an entry we haven't ever seen, so it is safe to discard the log.

The last `entryID` in the log can have an older term than `unstable.term` (which is `LogMark.Term`). Say, if leader @ term 5 replicated a log `[2 2 3 3]` to us with the latest append, the last entry term is 3, but the `LogMark.Term` is 5.

It is safe to discard the log if one of of the following is true, for the `entryID`@`term` snapshot:
- We don't have `entryID` in our log. This means our log is stale compared to the committed state, and we can replace it.
- `entryID.index` >= our `lastIndex`. Same reason.